fixed regex of STLLoader.js

cnspaha opened the pull request:
There was en failure in the regexpresion in the stl loader. Should be fine now.
Can be reproduced with the attached STL.

[test.zip](https://github.com/mrdoob/three.js/files/973580/test.zip)

cnspaha committed:
Add dispose function to VREffect. Otherwise it causes a memory leak if it is instantiated several times -> always add eventlistener on window

cnspaha committed:
Revert "Add dispose function to VREffect. Otherwise it causes a memory leak if it is instantiated several times -> always add eventlistener on window"

This reverts commit 05606328ecabd491876e37ed18dd598c74db1d6b.

cnspaha committed:
Make TransformControls also be usable with the edge browser on a microsoft surface

cnspaha committed:
set scope.axis before calling scope.dispatchEvent

cnspaha committed:
Merge branch 'transformcontrolsMSEdge'

cnspaha committed:
added uniform to transpose the values of red with cyan

uniform _invertRedCyan == false -> object appears IN the screen
uniform _invertRedCyan == true  -> object appears IN FRONT of the screen

cnspaha committed:
Merge branch 'master' of https://github.com/cadenasgmbh/three.js

cnspaha committed:
fixed regex of STLLoader.js

cnspaha committed:
Merge branch 'stlloader-regex-fix'

jostschmithals commented on examples/js/loaders/STLLoader.js:
```diff
@@ -222,7 +222,7 @@ THREE.STLLoader.prototype = {
 			}
 
-			patternVertex = /vertex[\s]+([\-+]?[0-9]+\.?[0-9]*([eE][\-+]?[0-9]+)?)+[\s]+([\-+]?[0-9]*\.?[0-9]+([eE][\-+]?[0-9]+)?)+[\s]+([\-+]?[0-9]*\.?[0-9]+([eE][\-+]?[0-9]+)?)+/g;
+			patternVertex = /vertex[\s]+([\-+]?[0-9]+\.?[0-9]*([eE][\-+]?[0-9]+)?)+[\s]+([\-+]?[0-9]*\.?[0-9]*([eE][\-+]?[0-9]+)?)+[\s]+([\-+]?[0-9]*\.?[0-9]*([eE][\-+]?[0-9]+)?)+/g;
```
I think these regex changes make sense.

On this occasion some additional questions: 

1. The first position marked in my screenshot (where I added line breaks only for readability) is different in the first line from the second and third line's version. If writing a float like .3 instead of 0.3 is allowed in the STL file format, it should also be `*` in this place.

2. I wonder why there are `+` signs at the end of each line? I think they should be removed?

3. It would be simpler and less error-prone to write only one of these "lines" (in parentheses) and add `{3}`.
![regex](https://cloud.githubusercontent.com/assets/5700294/25666912/2e75693a-3022-11e7-938a-7cbfec17150b.JPG)

The same situation with the normals pattern.

cnspaha replied:
1. You are right, but if we would just change that to "*"  then it would be also possible to match only a dot (".") as valid "number" which is of course wrong. But yeah .3 should be also allowed but that means that we would have to lookahead

2.  you are right

3. Also correct 

Here another regex which includes your points:
vertex([\s]+([\-+]?[0-9]*\.?[0-9]*([eE][\-+]?[0-9]+)?)){3}

The problem with this is that "vertex . 0.1 -4." would also be valid because of the reason mentioned in point 1. But haven't worked with lookahead in regex yet so if you a into it, please feel free to contribute ;)

jostschmithals replied:
1. It is clear that the regex could be more restrictive, but I guess this is not relevant in this case? On the other hand the regex should not be too restrictive, and it should not be different for the x coordinate and the other two.

2. This was only a question of mine  - I have no experience with STL, so I don't know whether there could be situations where the final `+` signs are used, though they are unnecessary in a file like your test.stl.

3. Now I saw that the simplification with {3} wouldn’t work, because the values of the first, third and fifth parentheses are used later in the code. Sorry, that I didn’t consider the code context yesterday.

mrdoob commented:
Do you mind clarifying the text that the old regex isn't finding?

cnspaha commented:
@mrdoob 
yeah sure for example this line
vertex .0 1.0 1.0
the regex wouldn't match. But ".0" is a valid number-
but will soon commit a lookahead regex which should be able to handle it properbly, considering the points @jostschmithals  mentioned. stay tuned ;)

jostschmithals commented:
@mrdoob You asked which kind of text pattern the “old regex” doesn’t recognize. If you meant the code of the current dev branch (which doesn’t work with @cnspaha‘s test.stl), the problem are numbers like `4.` (without a digit **after** the decimal point).

A number like `4.` is not recognized by this part of the regex: `[0-9]*\.?[0-9]+` , since `+` expects at least one digit.

So the parsing goes wrong, at least if such a number (which has a decimal point, but no decimals) represents the `x` or `y` coordinate of the vertex.
 
A little curious at first glance: For `z` the pattern works though. This is more or less “accidentally”, because `[0-9]*` and `\.?` will be skipped, `[0-9]+` matches the digit on the **left** of the point, and the point (as the last sign) is left out (which doesn't matter).

cnspaha commented:
The new regex with lookahead:
vertex[\s]+([+\-]?(?:\d+\.\d+|\d+\.|\d+|\.\d+)(?:[eE][+\-]?\d+)?)[\s]+([+\-]?(?:\d+\.\d+|\d+\.|\d+|\.\d+)(?:[eE][+\-]?\d+)?)[\s]+([+\-]?(?:\d+\.\d+|\d+\.|\d+|\.\d+)(?:[eE][+\-]?\d+)?)

Should be fine now.
You can test it on https://regex101.com/r/OlvqAf/6 , there I already put some test data into.
If you are fine with it, I will commit it on the branch.

jostschmithals commented:
@cnspaha No doubt, your pattern works very well in all cases of your impressive test site - and it doesn’t match vertex data which contain a solitary dot instead of a valid number (this is what you wanted to achieve, if I understood it correctly?).

Nevertheless I have two questions:

1.
It’s not yet clear to me what the advantage of your more complex pattern would be in real life(?)

If I replace `0. ` by `.` in the z component of the third vertex of your `test.stl` and load it with “my” pattern …

`vertex[\s]+([\-+]?[0-9]*\.?[0-9]*([eE][\-+]?[0-9]+)?)[\s]+([\-+]?[0-9]*\.?[0-9]*([eE][\-+]?[0-9]+)?)[\s]+([\-+]?[0-9]*\.?[0-9]*([eE][\-+]?[0-9]+)?) `       (normals pattern accordingly)

… only the according first face is not displayed (and an error message is thrown), but all the other faces look fine. I think this is a reasonable output, since only data of one vertex are not valid.

If I try the same with your pattern - using the value of the 1., 2., and 3. capturing group instead of the 1., 3. and 5., since lookarounds are not included in the count -, the whole geometry is destroyed. This is because the invalid vertex isn’t parsed at all, which causes the following vertices being combined wrongly when building the faces.
 
Hence some extra logic would be needed to handle this.

2.
I don't know: How much does the parsing time increase by your lookaround, if the user loads a large model or many models? 

I have no experience with this, and with your very small testfile I was not able to evaluate it, since the statistical scattering is very high.

If the performance loss turns out to be significant for larger models, it would perhaps be better to simply assume that the model file doesn’t contain solitary dots.

cnspaha commented:
1. `.`  isn't a valid number so it shouldn't be allowed. My regex doesn't allow it.
of course you will have to use the 1., 2. and 3. capturing group instead of 1., 3., and 5..
I just put my regex into the STLLoader.js and also adjusted the capturing groups. 
That is how it looks now:

`	
	
	text = result[ 0 ];
	patternNormal = /normal[\s]+([+-]?(?:\d+.\d+|\d+.|\d+|.\d+)(?:[eE][+-]?\d+)?)[\s]+([+-]?(?:\d+.\d+|\d+.|\d+|.\d+)(?:[eE][+-]?\d+)?)[\s]+([+-]?(?:\d+.\d+|\d+.|\d+|.\d+)(?:[eE][+-]?\d+)?)/g;

	while ( ( result = patternNormal.exec( text ) ) !== null ) {

		normal.x = parseFloat( result[ 1 ] );
		normal.y = parseFloat( result[ 2 ] );
		normal.z = parseFloat( result[ 3 ] );

	}

	patternVertex = /vertex[\s]+([+-]?(?:\d+.\d+|\d+.|\d+|.\d+)(?:[eE][+-]?\d+)?)[\s]+([+-]?(?:\d+.\d+|\d+.|\d+|.\d+)(?:[eE][+-]?\d+)?)[\s]+([+-]?(?:\d+.\d+|\d+.|\d+|.\d+)(?:[eE][+-]?\d+)?)/g;

	while ( ( result = patternVertex.exec( text ) ) !== null ) {

		vertices.push( parseFloat( result[ 1 ] ), parseFloat( result[ 2 ] ), parseFloat( result[ 3 ] ) );
		normals.push( normal.x, normal.y, normal.z );

	}
		`
I just tested it with following code:

    var loader = new THREE.STLLoader();
    var t0 = window.performance.now();
    var geom = loader.parse(data);
    var t1 = window.performance.now();
    console.log(Math.round(t1 - t0));

and took the times with the previous regex and with the lookaround regex.
The STL file I used is ~38MB big so I assume the size for testing should be fine.
Against my assumption, that the lookaround regex is slower, the result suprised me.. it seems like that the lookaround regex is a bit faster (~4%) than the old one...
Here you can see my result. 
![batman_stl](https://cloud.githubusercontent.com/assets/22448974/25740798/50041788-3188-11e7-850d-63d9a45b873c.PNG)

jostschmithals commented:
That’s indeed surprising - but I can reproduce it (on a much slower machine): a slight advantage for the lookaround version in Chrome, and no significant difference in Firefox.

Therefore it seems that we can ignore the performance aspect. But I think if you want to use the lookaround pattern, you will have to handle the regex’s result differently from the current code as well, according to what I wrote in my previous post (1.), so that not the whole geometry will be destroyed if only one number is not valid.

BTW: I compared the durations of the while loops directly, to eliminate other effects – there seems to be a performance issue which is much more important; but I will make a separate post for this, because it is a totally different thing.

cnspaha commented:
The part where you have to handle the results differently are that u have to take the parts [1], [2], [3] instead of [1], [3], [5]. Is there any other part?
btw if an number in the STL is invalid then of course errors occur. But something like "fixing" a broken part is a completely different story ;)

so can we agree, that we use the code from my previous post?

jostschmithals commented:
Apparently it has not become clear what I meant. Therefore here two screenshots.

Your goal was that the regex doesn't match solitary dots. So I made in your test.stl the first triangle invalid using a single dot as z component of the third vertex.

With "my" pattern (from your PR, with the red marked changes) ... 
```
[\s]+([\-+]?[0-9]*\.?[0-9]*([eE][\-+]?[0-9]+)?)
[\s]+([\-+]?[0-9]*\.?[0-9]*([eE][\-+]?[0-9]+)?)
[\s]+([\-+]?[0-9]*\.?[0-9]*([eE][\-+]?[0-9]+)?)
```
...you can directly see which triangle is the invalid one and easily fix the defect:

![stl1](https://cloud.githubusercontent.com/assets/5700294/25797870/4f0ee624-33df-11e7-84d3-babd8f131b64.JPG)

With your pattern all looks ruined, so that it would not be obvious at all that there is actually only one little defect, and where to search for it:

![stl2](https://cloud.githubusercontent.com/assets/5700294/25797910/769d09dc-33df-11e7-9c2f-d91ffd57a9df.JPG)

donmccurdy commented:
IMO this could be cleaned up a little for easier reading:

```js
var patternNormal = new RegExp(''
    + 'normal'
    + /[\s]+([\-+]?[0-9]+\.?[0-9]*([eE][\-+]?[0-9]+)?)+/.source
    + /[\s]+([\-+]?[0-9]*\.?[0-9]*([eE][\-+]?[0-9]+)?)+/.source
    + /[\s]+([\-+]?[0-9]*\.?[0-9]*([eE][\-+]?[0-9]+)?)+/.source
);
```

^although perhaps best to put that outside the loop.

cnspaha commented:
@jostschmithals 
I get your point. With your method it is indeed easier to spot the wrong triangle and this is indeed nice but my pattern is the more correct one even though it is more prone to wrong models. but IMO you should be able to spot bad models already in your modeling software and that shouldn't be the part of three.js. So I would want to keep my pattern instead of yours. 

@mrdoob 
since you are the one to have the last word, what is your opinion about this?

@donmccurdy 
yeah, I had also thought about putting them outside the loop and clean them up for easier reading and I will do it with one of the next commits, But first I would like to know if my pattern will be accepted before doing anything else.

jostschmithals commented:
@cnspaha 

> IMO you should be able to spot bad models already in your modeling software and that shouldn't be the part of three.js

OK (if it is an own model). But If you assume that the model has no solitary dots I don't understand why you need your lookarond pattern(?)

I try to summarize from my point of view:

**If a file is valid** (in the sense that it contains no solitary dots instead of numbers), it is correctly parsed with the simpler regex, and with your more complicated one as well.

**If a file is not valid**, since it contains a solitary dot instead of a number, your pattern removes the wrong vertex (which is theoretically better), but it destroys the following faces (which is worse in practice). In this case the simpler regex ignores the file’s imperfection (which is theoretically worse), but it leaves the valid faces untouched (which is better in practice).

Hence in sum I would still prefer the simpler one with the slight modifications suggested here (my red marks and @donmccurdy’s comment). 

But of course @mrdoob has to decide how to go on.

mrdoob commented:
How do you end up with invalid files? If I have to chose, I would just support valid files.

cnspaha commented:
@jostschmithals 
you are right with the invalid file, that your pattern destroys the part less. But your pattern also sees a solitary dot as valid, which is IMO just wrong.
So I would rather throw an error or so when my pattern does not match, to tell the user/developer that the stl file is invalid.

jostschmithals commented:
@cnspaha So it seems that we are in agreement about this  :-)

In my recent comments I only wanted to point out what would happen if you simply replace the old regex by your lookaround pattern (adjusting the numbers of the capturing groups), but don’t react to the case in which a file is invalid.

mrdoob commented:
So is this ready to merge then? 😅

cnspaha commented:
@jostschmithals 
okay thanks ;)

@mrdoob 
not quite yet ;) 
have to commit the stuff first, will do it in the next days :)

cnspaha committed:
fix: corrected regexp for vertices and normals in stl files
mod: throw error when there isn't exactly one valid normal and three valid vertices per face

cnspaha commented:
@mrdoob 
sorry for the delay, was so caught up in work, so there was almost not time for this..
but should be fine now. Also added throwing of an error if something is wrong with a normal or the vertices. What message we should throw exactly when an error occurs is of course something we can discuss ;)

mrdoob commented:
@jostschmithals looks good?

jostschmithals commented:
@mrdoob @cnspaha Will have a closer look at it tomorrow.

jostschmithals commented on examples/js/loaders/STLLoader.js:
```diff
@@ -200,37 +200,63 @@ THREE.STLLoader.prototype = {
+			}
+			faceCounter++;
 		}
 
```
Only some little notes on this file:

1.	`length` (in the past defined in line 203) could be removed since it is never used 
2.	You are throwing a string (which is possible in JavaScript), but common practice in three.js seems to be to throw a `new Error`
3.	Small typo (missing “t”) in isn\’t (2x)
4.	Shouldn’t it be better “Each face has” instead of “Every face have” (2x)?
5.	Perhaps you should check the code with http://zz85.github.io/mrdoobapproves/, so that it matches @mrdoob’s code style guide (there are some missing spaces and some empty lines with whitespace)

And one question: two other files that have no relation to this issue are indicated as changed, too. I don’t know what this means and if it could be a problem.

mrdoob replied:
> And one question: two other files that have no relation to this issue are indicated as changed, too. I don’t know what this means and if it could be a problem.

True. Seems like this PR got a bit corrupted.

cnspaha committed:
created fix for stlloader discussed at https://github.com/mrdoob/three.js/pull/11272, but messed up other branch

cnspaha commented:
ah sorry, really messed this branch up... :/ 
just created a new CLEAN branch with which I will provide a PR. so this PR here should be closed... and ignored ;)
@jostschmithals  I put your 5 points into account in my new PR
@mrdoob  please close this one and approve the other one I will PR ;)
